R/tinytheme.R

```diff
@@ -0,0 +1,153 @@
+#' Set or Reset Plot Themes for `tinyplot`
+#'
+#' @description
+#' The `tinytheme` function sets or resets the theme for plots created with
+#' `tinyplot`. Themes control the appearance of plots, such as text alignment,
+#' font styles, and axis labels. By default, a "bw" theme is available.
+#'
+#' Warning: `tinytheme()` shuts down all open graphical devices.
+#'
+#' @param theme A character string specifying the name of the theme to apply.
+#'   - `"default"`
+#'   - `"bw"`
+#'   - `"classic"`
+#'   - `"dark"`
+#'   - `"ipsum"`
+#'   - `"minimal"`
+#' @param ... Named arguments to override specific theme settings. These arguments are
+#'   passed to `tpar()` and take precedence over the predefined settings in the selected
+#'   theme.
+#'
+#' @details
+#' Sets a list of graphical parameters using `tpar()`
+#'
+#' To reset the theme to default settings (no customization), call `tinytheme()` without arguments.
+#'
+#' The `"custom"` theme sets parameters stored in a list in a global option. See examples below.
+#'
+#' @return The function returns nothing. It is called for its side effects.
+#'
+#' @examples
+#' # Set a theme
+#' tinytheme("bw")
+#' tinyplot(mpg ~ hp | factor(am), data = mtcars)
+#'
+#' # Customize the theme by overriding default settings
+#' tinytheme("bw", fg = "blue", font.main = 2)
+#' tinyplot(mpg ~ hp | factor(am), data = mtcars, main = "Hello World!")
+#'
+#' # Reset the theme
+#' tinytheme()
+#' tinyplot(mpg ~ hp | factor(am), data = mtcars)
+#'
+#' # Custom theme
+#' tinytheme("default", font.main = 3, col = "red")
+#' tinyplot(mpg ~ hp | factor(am), data = mtcars)
+#'
+#' @export
+tinytheme = function(theme = "default", ...) {
+  # in interactive sessions, we close the graphics device to reset parameters
+  if (interactive()) {
+    grDevices::graphics.off()
+  } else {
+    par_first = get_saved_par("first")
+    if (!is.null(par_first)) {
+      do.call(tpar, par_first)
+    } else {
+      set_saved_par("first", par())
+    }
+  }
+
+  # in notebooks, we don't want to close the device because no image.
+  # init_tpar() tries to be smarkt, but may fail.
+  init_tpar(rm_hook = TRUE)
+
+  assert_choice(theme, c("default", sort(c("bw", "classic", "dark", "ipsum", "minimal"))))
+
+  settings = switch(theme,
+    "default" = theme_default,
+    "bw" = theme_bw,
+    "classic" = theme_classic,
+    "dark" = theme_dark,
+    "ipsum" = theme_ipsum,
+    "minimal" = theme_minimal
+  )
+
+  dots = list(...)
+  for (n in names(dots)) {
+    settings[[n]] = dots[[n]]
+  }
+
+  if (length(settings) > 0) {
+    tpar(settings, hook = TRUE)
+  }
+
+  return(invisible(NULL))
+}
+
+
+theme_default = list()
+
+theme_bw = list(
+  tinytheme = "bw",
+  adj = 0.5,
+  adj.main = 0,
+  adj.sub = 0,
```

Comment:
I'm not sure exactly how to do this. What if we call `tinytheme()` when there is no active graphical device? I think this requires some experimentation, and probably some logic in `tinyplot()` itself rather than in the themes. I'm not sure and don't have a good understanding of how margins and plot areas are drawn.

Reply:
Margin behaviour is retrieved/governed in `draw_legend.R`, starting here. https://github.com/grantmcdermott/tinyplot/blob/7ee0725bbb1d5b7788f055a5529d15ac194bd60e/R/draw_legend.R#L200-L204

(Reminder: `draw_legend.R` is where we first open the new plot window, since we need to manipulate the plot margins for correct legend placement before we do anything else.)

With that in mind, one option is to introduce a new `mar_adj` parameter that could live in `tpar()`, and is thus accessible to `tinytheme()` too, which applies an adjustment factor to the `omar` variable that we create/call in the code link above. By default this would be `mar_adj = c(0,0,0,0)`, corresponding to zero `c(bottom, left, top, right)` adjustment.

For themes where `sub` is on moved to the top, we could do `mar_adj = c(1,0,0,0)`, i.e. subtract 1 line from the bottom margin.

I can test this out and push the change if it looks good.